### PR TITLE
dotproducttrans fix (closes #1834)

### LINKED ISSUE #1834: dotproduct transformation limitation
The dotproduct transformation only works if the inner dimension of the argument is a ':'. This is too restrictive as it could have bounds as long as they are the full range e.g.

```wind(1:3) in DOT_PRODUCT(basis_w1(:,df1,qp1,qp2), wind(1:3)) where wind is an array of size 3```

### PR DESCRIPTION
**Pull request by rupertford**
Adds support for explicit array range declarations to the dotproducttrans transformation.

Required for PSyAD with PR #1825.

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for first review from @arporter or @sergisiso

File: src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py [hunk 1]
```diff
@@ -537,3 +551,31 @@ def test_apply_multi_different_line(tmpdir, fortran_reader, fortran_writer):
         "  enddo\n"
         "  result = result + c * res_dot_product_1\n")
     check_trans(code, expected, fortran_reader, fortran_writer, tmpdir)
+
+
+def test_apply_explicit_range(fortran_reader, fortran_writer, tmpdir):
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py:556]: This test fails for me when compilation is enabled:

    ----------------------------- Captured stderr call -----------------------------
    Compiling: gfortran generated-8788527625436.f90 -c generated-8788527625436.f90:3:32:

        3 |   real, dimension(:) :: basis_w1
          |                                1
    Error: Array ‘basis_w1’ at (1) cannot have a deferred shape
    
    =========================== short test summary info ============================
    FAILED psyir/transformations/intrinsics/dotproduct2code_trans_test.py::test_apply_explicit_range

**Review by arporter — CHANGES_REQUESTED** [review]
Nice improvement.
There's some tidying needed and I would expect to see the explicit tests of `_is_upper/lower_bound` extended to check the new functionality. Please could you do this.
The new test fails when compilation is enabled.
I'll check with pylint next time.

File: src/psyclone/psyir/nodes/array_mixin.py [hunk 2]
```diff
@@ -127,6 +126,8 @@ def is_lower_bound(self, index):
         'index' matches the specified array index. Otherwise False is
         returned.
 
+        **** RF: Or if the starting value of the range node is an integer that matches the starting value of the declaration. ***
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/nodes/array_mixin.py:129]: Oops, reformatted.

File: src/psyclone/psyir/nodes/array_mixin.py [hunk 3]
```diff
@@ -140,11 +141,27 @@ def is_lower_bound(self, index):
         '''
         self._validate_index(index)
 
+
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/nodes/array_mixin.py:144]: Oops.

File: src/psyclone/psyir/nodes/array_mixin.py [hunk 4]
```diff
@@ -140,11 +141,27 @@ def is_lower_bound(self, index):
...
 
         lower = array_dimension.start
+
+        if isinstance(lower, Literal):
+            try:
+                symbol = self.scope.symbol_table.lookup(self.name)
+                datatype = symbol.datatype
+                shape = datatype.shape
+                array_bounds = shape[index]
+                if isinstance(array_bounds, ArrayType.ArrayBounds) \
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/nodes/array_mixin.py:157]: OK.

File: src/psyclone/psyir/nodes/array_mixin.py [hunk 5]
```diff
@@ -140,11 +141,27 @@ def is_lower_bound(self, index):
...
+                   and isinstance(array_bounds.lower, Literal):
+                    if lower.value == array_bounds.lower.value:
+                        return True
+            except (KeyError, SymbolError):
+                # Can't find symbol declaration
+                pass
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/nodes/array_mixin.py:163]: Added.

File: src/psyclone/psyir/nodes/array_mixin.py [hunk 6]
```diff
@@ -187,6 +204,21 @@ def is_upper_bound(self, index):
             return False
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/nodes/array_mixin.py:204]: Extended.

File: src/psyclone/psyir/nodes/array_mixin.py [hunk 7]
```diff
@@ -187,6 +204,21 @@ def is_upper_bound(self, index):
...
 
         upper = array_dimension.stop
+
+        if isinstance(upper, Literal):
+            try:
+                symbol = self.scope.symbol_table.lookup(self.name)
+                datatype = symbol.datatype
+                shape = datatype.shape
+                array_bounds = shape[index]
+                if isinstance(array_bounds, ArrayType.ArrayBounds) and \
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/nodes/array_mixin.py:214]: Replaced.

File: src/psyclone/psyir/nodes/array_mixin.py [hunk 8]
```diff
@@ -187,6 +204,21 @@ def is_upper_bound(self, index):
...
+                   isinstance(array_bounds.upper, Literal):
+                    if upper.value == array_bounds.upper.value:
+                        return True
+            except (KeyError, SymbolError):
+                # Can't find symbol declaration
+                pass
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/nodes/array_mixin.py:220]: Done.

File: src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py [hunk 9]
```diff
@@ -97,6 +97,7 @@ def check_trans(code, expected, fortran_reader, fortran_writer, tmpdir):
         if bin_op.operator == BinaryOperation.Operator.DOT_PRODUCT:
             trans.apply(bin_op)
     result = fortran_writer(psyir)
+    print(result)
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py:100]: Oops.

File: src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py [hunk 10]
```diff
@@ -286,6 +287,19 @@ def test_validate_array_full_slice(fortran_reader):
     check_validate(code, expected, fortran_reader)
 
 
+#def test_validate_array_full_slice_2(fortran_reader):
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py:290]: Oops.

File: src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py:556]: Oops. Fixed.

**Comment by rupertford** [comment]
I've added `is_lower_bound()` and `is_upper_bound()` tests to `array_mixin_test.py`. Hopefully that is what you were after.

**Comment by rupertford** [comment]
Ready for next review from @arporter

File: src/psyclone/psyir/nodes/array_mixin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/array_mixin.py:162]: I meant instead of the `pass` :-)

File: src/psyclone/psyir/nodes/array_mixin.py [hunk 11]
```diff
@@ -182,9 +182,11 @@ def is_lower_bound(self, index):
     def is_upper_bound(self, index):
         '''Returns True if the specified array index contains a Range node
         which has a stopping value given by the 'UBOUND(name,index)'
-        intrinsic where 'name' is the name of the current ArrayReference and
-        'index' matches the specified array index. Otherwise False is
-        returned.
+        intrinsic where 'name' is the name of the current
+        ArrayReference and 'index' matches the specified array index.
+        Also returns True if the starting value of the range node is
```
Comment (arporter) [inline src/psyclone/psyir/nodes/array_mixin.py:187]: s/starting/stopping/ and again below.

File: src/psyclone/psyir/nodes/array_mixin.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/array_mixin.py:222]: Again, I meant instead of `pass`?

**Review by arporter — CHANGES_REQUESTED** [review]
Almost there now. Just a few tiny changes and please could you bring to master.

File: src/psyclone/psyir/nodes/array_mixin.py (hunk 11 above)
Comment (rupertford) [reply src/psyclone/psyir/nodes/array_mixin.py:187]: Oops.

File: src/psyclone/psyir/nodes/array_mixin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/nodes/array_mixin.py:162]: In that case, no we can't, or at least we would need another return False. We may get a False by falling through the try without an exception so need to catch that as well.

**Comment by arporter** [comment]
Not sure what's happened here @rupertford but you've suddenly got some failing tests :-(

**Comment by rupertford** [comment]
Broken tests should now be fixed. Apologies.

**Review by arporter — APPROVED** [review]
Tests and examples are now all OK again.
Will proceed to merge...

**Comment by arporter** [comment]
There's a link problem in sphinx link check but that's not the fault of this PR.

**Pull request merged**